dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java

```diff
@@ -231,119 +229,156 @@ protected void populateTable( AnalyticsTableUpdateParams params, AnalyticsTableP
             }
         }
 
-        List<OrganisationUnitLevel> levels =
-            organisationUnitService.getFilledOrganisationUnitLevels();
+        // Wrapper object for Program, required to avoid errors with Hibernate and non thread-safe Session
 
-        List<OrganisationUnitGroupSet> orgUnitGroupSets =
-            idObjectManager.getDataDimensionsNoAcl( OrganisationUnitGroupSet.class );
+        DetachedProgram detachedProgram = new DetachedProgram(program);
```

Comment:
Ideally we would avoid having a `DetachedProgram` class. There are ways to attach background jobs to the Hibernate session.

Reply:
I have tried using `Hibernate.initialize( program )` but I still stumble in the same concurrent session access error. I see two alternatives: 

1. remove Futures and access the `Program` object from  a single thread
2. Use this pretty awful approach (which works, but..)
```
 private void init( Program program )
    {
        program.getDataElements().size();
        program.getDataElementsWithLegendSet().size();
        program.getNonConfidentialTrackedEntityAttributes().size();
        program.getNonConfidentialTrackedEntityAttributesWithLegendSet().size();
    }
```